<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Scramble a matrix of genotype data — mat_scramble • gscramble</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Scramble a matrix of genotype data — mat_scramble"><meta property="og:description" content="This function assumes that M is a matrix with L rows (number of markers) and
2 * N (N = number of individuals) columns.
There are two ways that the data might be permuted.  In the first,
obtained with preserve_haplotypes = FALSE,
the position of missing data within the matrix is held constant, but all
non-missing sites within a row (i.e. all gene copies at a locus) get
scrambled amongst the samples.  In the second way, just the columns are
permuted.  This preserves haplotypes in the data, if there are any.
The second approach should only be used if haplotypes are inferred in
the individuals."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gscramble</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.1.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/gscramble-tutorial.html">A first tutorial for 'gscramble'</a>
    </li>
    <li>
      <a href="../articles/about-createGSP.html">Pedigrees Available from create_GSP()</a>
    </li>
    <li>
      <a href="../articles/permutation-options.html">Permutation Options in 'gscramble'</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/eriqande/gscramble/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Scramble a matrix of genotype data</h1>
    <small class="dont-index">Source: <a href="https://github.com/eriqande/gscramble/blob/HEAD/R/mat_scramble.R" class="external-link"><code>R/mat_scramble.R</code></a></small>
    <div class="hidden name"><code>mat_scramble.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function assumes that M is a matrix with L rows (number of markers) and
2 * N (N = number of individuals) columns.
There are two ways that the data might be permuted.  In the first,
obtained with <code>preserve_haplotypes = FALSE</code>,
the position of missing data within the matrix is held constant, but all
non-missing sites within a row (i.e. all gene copies at a locus) get
scrambled amongst the samples.  In the second way, just the columns are
permuted.  This preserves haplotypes in the data, if there are any.
The second approach should only be used if haplotypes are inferred in
the individuals.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">mat_scramble</span><span class="op">(</span></span>
<span>  <span class="va">M</span>,</span>
<span>  preserve_haplotypes <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  row_groups <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  preserve_individuals <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>M</dt>
<dd><p>a matrix with L rows (number of markers) and 2 * N columns
where N is the number of individuals. Missing data must be coded
as NA</p></dd>


<dt>preserve_haplotypes</dt>
<dd><p>logical indicating whether the haplotypes
set to be TRUE</p></dd>


<dt>row_groups</dt>
<dd><p>if not NULL must be a list of indexes of adjacent rows
that are all in the same groups.  For example: <code>list(1:10, 11:15, 16:30)</code>.
They should be in order and complete.
In practice, these should correspond to the indexes of markers on different
chromosomes.</p></dd>


<dt>preserve_individuals</dt>
<dd><p>logical indicating whether the genes within
each individual should stay togeter.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>This function returns a matrix of the same dimensions and storage.mode
as the input, <code>M</code>; however the elements have been permuted according to the
options specified by the users.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>There is now an additional way of permuting:  if
<code>preserve_individuals = TRUE</code>, then entire individuals are permuted.
If <code>preserve_haplotypes = FALSE</code>, then the gene copies at each locus
are randomly ordered within each individual before permuating them.
If <code>preserve_haplotypes = TRUE</code> then that initial permutation is not
done.  This should only be done if the individuals are phased and that
phasing is represented in how the genotypes are stored in the matrix.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># make a matrix with alleles named as I.M.g, where I is individual</span></span></span>
<span class="r-in"><span><span class="co"># number, M is marker number, and g is either "a" or "b" depending</span></span></span>
<span class="r-in"><span><span class="co"># on which gene copy in the diploid it is.  4 indivs and 7 markers...</span></span></span>
<span class="r-in"><span><span class="va">Mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span></span>
<span class="r-in"><span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>   <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, each <span class="op">=</span> <span class="fl">7</span> <span class="op">*</span> <span class="fl">2</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>   <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">7</span>, <span class="fl">4</span> <span class="op">*</span> <span class="fl">2</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>   <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"b"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">7</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>   sep <span class="op">=</span> <span class="st">"."</span></span></span>
<span class="r-in"><span> <span class="op">)</span>,</span></span>
<span class="r-in"><span> nrow <span class="op">=</span> <span class="fl">7</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># without preserving haplotypes</span></span></span>
<span class="r-in"><span><span class="va">S1</span> <span class="op">&lt;-</span> <span class="fu">mat_scramble</span><span class="op">(</span><span class="va">Mat</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># preserving haplotypes with markers 1-7 all on one chromosome</span></span></span>
<span class="r-in"><span><span class="va">S2</span> <span class="op">&lt;-</span> <span class="fu">mat_scramble</span><span class="op">(</span><span class="va">Mat</span>, preserve_haplotypes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># preserving haplotypes with markers 1-3 on one chromosome and 4-7 on another</span></span></span>
<span class="r-in"><span><span class="va">S3</span> <span class="op">&lt;-</span> <span class="fu">mat_scramble</span><span class="op">(</span><span class="va">Mat</span>, row_groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">4</span><span class="op">:</span><span class="fl">7</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># preserving individuals, but not haplotypes, with two chromosomes</span></span></span>
<span class="r-in"><span><span class="va">S4</span> <span class="op">&lt;-</span> <span class="fu">mat_scramble</span><span class="op">(</span><span class="va">Mat</span>, row_groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">4</span><span class="op">:</span><span class="fl">7</span><span class="op">)</span>, preserve_individuals <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># preserving individuals by chromosome, but not haplotypes, with two chromosomes</span></span></span>
<span class="r-in"><span><span class="va">S5</span> <span class="op">&lt;-</span> <span class="fu">mat_scramble</span><span class="op">(</span><span class="va">Mat</span>, row_groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">4</span><span class="op">:</span><span class="fl">7</span><span class="op">)</span>, preserve_individuals <span class="op">=</span> <span class="st">"BY_CHROM"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># preserving individuals by chromosome, and preserving haplotypes, with two chromosomes</span></span></span>
<span class="r-in"><span><span class="va">S6</span> <span class="op">&lt;-</span> <span class="fu">mat_scramble</span><span class="op">(</span><span class="va">Mat</span>, row_groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">4</span><span class="op">:</span><span class="fl">7</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 preserve_individuals <span class="op">=</span> <span class="st">"BY_CHROM"</span>, preserve_haplotypes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Eric C. Anderson, Rachael M. Giglio, Matt G. DeSaix, Timothy J. Smyser.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

