<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Map alleles from scrambled founders to the sampled segments from a GSP. — segments2markers • gscramble</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Map alleles from scrambled founders to the sampled segments from a GSP. — segments2markers"><meta property="og:description" content="Map alleles from scrambled founders to the sampled segments from a GSP."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gscramble</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/gscramble-tutorial.html">A first tutorial for 'gscramble'</a>
    </li>
    <li>
      <a href="../articles/about-createGSP.html">Pedigrees Available from create_GSP()</a>
    </li>
    <li>
      <a href="../articles/permutation-options.html">Permutation Options in 'gscramble'</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/eriqande/gscramble/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Map alleles from scrambled founders to the sampled segments from a GSP.</h1>
    <small class="dont-index">Source: <a href="https://github.com/eriqande/gscramble/blob/HEAD/R/segments2markers.R" class="external-link"><code>R/segments2markers.R</code></a></small>
    <div class="hidden name"><code>segments2markers.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Map alleles from scrambled founders to the sampled segments from a GSP.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">segments2markers</span><span class="op">(</span></span>
<span>  <span class="va">Segs</span>,</span>
<span>  <span class="va">Im</span>,</span>
<span>  <span class="va">Mm</span>,</span>
<span>  <span class="va">G</span>,</span>
<span>  preserve_haplotypes <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  preserve_individuals <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>Segs</dt>
<dd><p>the simulated segments. A tibble like that returned from
<code><a href="segregate.html">segregate()</a></code>.</p></dd>


<dt>Im</dt>
<dd><p>the individual meta data, like that in <code><a href="I_meta.html">I_meta</a></code>. A tibble with
columns <code>group</code> and <code>indiv</code>.</p></dd>


<dt>Mm</dt>
<dd><p>the marker meta data formatted like that in <code><a href="M_meta.html">M_meta</a></code>. A tibble
with columns <code>chrom</code>, <code>pos</code>, and <code>variant_id</code>.</p></dd>


<dt>G</dt>
<dd><p>the marker genotype data as a matrix like <code><a href="Geno.html">Geno</a></code>.  This is
a character matrix.  Each row is an individual, and each pair of columns are the
alleles at a locus.  Thus it is N x 2L where N is the number of individuals
and L is the number of markers.</p></dd>


<dt>preserve_haplotypes</dt>
<dd><p>If TRUE then the Geno data is assumed phased
(first allele at an individual on one haplotype and second allele on the
other) and those haplotypes are preserved in this permutation of
genomic material amongst the founders.</p></dd>


<dt>preserve_individuals</dt>
<dd><p>If TRUE then whole individuals are permuted
around the data set and the two gene copies at each locus are randomly
permuted within each individual.  If <code>preserve_individuals = "BY_CHROM"</code>,
then the the two copies of each chromosome in an individual are permuted
together.  Thus a permuted individual may have two copies of one chromosome
from one individual, and two copies of another chromosome from a different
individual.  (If <code>preserve_haplotypes = TRUE</code> then
the gene copies are not permuted within individuals. You should only ever
use <code>preserve_haplotypes = TRUE</code> if you have phased data.)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list with three components:</p><ul><li><p><code>ret_geno</code>: A character matrix where each row is an individual and each pair of
columns are the alleles at a locus, thus it is N x 2L where N is the number of
individuals and L is the number of markers.</p></li>
<li><p><code>ret_ids</code>: A tibble providing the individual meta data with columns <code>groups</code> and <code>indiv</code>.</p></li>
<li><p><code>hyb_Qs</code>: A tibble of the admixture Q values.</p></li>
</ul></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">#### First, get input segments for the function ####</span></span></span>
<span class="r-in"><span><span class="co"># We construct an example here where we will request segregation</span></span></span>
<span class="r-in"><span><span class="co"># down a GSP with two F1s and F1B backcrosses between two hypothetical</span></span></span>
<span class="r-in"><span><span class="co"># populations, A and B.</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gsp_f1f1b</span> <span class="op">&lt;-</span> <span class="fu"><a href="create_GSP.html">create_GSP</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, F1 <span class="op">=</span> <span class="cn">TRUE</span>, F1B <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># We will imagine that in our marker data there are three groups</span></span></span>
<span class="r-in"><span><span class="co"># labelled "Pop1", "Pop2", and "Pop3", and we want to create the F1Bs with backcrossing</span></span></span>
<span class="r-in"><span><span class="co"># only to Pop3.</span></span></span>
<span class="r-in"><span><span class="va">reppop</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    pop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Pop3"</span>, <span class="st">"Pop1"</span>, <span class="st">"Pop3"</span>, <span class="st">"Pop2"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># combine those into a request</span></span></span>
<span class="r-in"><span><span class="va">request</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>   gpp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">gsp_f1f1b</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>   reppop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">reppop</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># now segegate segments.  Explicitly pass the markers</span></span></span>
<span class="r-in"><span><span class="co"># in M_meta so that the order of the markers is set efficiently.</span></span></span>
<span class="r-in"><span><span class="va">segs</span> <span class="op">&lt;-</span> <span class="fu"><a href="segregate.html">segregate</a></span><span class="op">(</span><span class="va">request</span>, <span class="va">RecRates</span>, <span class="va">M_meta</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#### Now, use segs in an example with segments2markers() ####</span></span></span>
<span class="r-in"><span><span class="co"># this uses several package data objects that are there for examples</span></span></span>
<span class="r-in"><span><span class="co"># and illustration.</span></span></span>
<span class="r-in"><span><span class="va">s2m_result</span> <span class="op">&lt;-</span> <span class="fu">segments2markers</span><span class="op">(</span><span class="va">segs</span>, <span class="va">I_meta</span>, <span class="va">M_meta</span>, <span class="va">Geno</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Eric C. Anderson, Rachael M. Giglio, Matt G. DeSaix, Timothy J. Smyser.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

