<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A first tutorial for `gscramble` • gscramble</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="A first tutorial for `gscramble`">
<meta property="og:description" content="gscramble">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gscramble</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/about-createGSP.html">Pedigrees Available from create_GSP()</a>
    </li>
    <li>
      <a href="../articles/gscramble-tutorial.html">A first tutorial for `gscramble`</a>
    </li>
    <li>
      <a href="../articles/gscramble-tutorial_1.2.html">A first tutorial for `gscramble`</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>A first tutorial for <code>gscramble</code>
</h1>
            
      
      
      <div class="hidden name"><code>gscramble-tutorial_1.2.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">gscramble</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>
<span class="co">#&gt; ── <span style="font-weight: bold;">Attaching packages</span> ─────────────────────────────────────── tidyverse 1.3.1 ──</span>
<span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">ggplot2</span> 3.3.6     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">purrr  </span> 0.3.4</span>
<span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">tibble </span> 3.1.7     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">dplyr  </span> 1.0.9</span>
<span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">tidyr  </span> 1.2.0     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">stringr</span> 1.4.0</span>
<span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">readr  </span> 2.1.2     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">forcats</span> 0.5.1</span>
<span class="co">#&gt; ── <span style="font-weight: bold;">Conflicts</span> ────────────────────────────────────────── tidyverse_conflicts() ──</span>
<span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">dplyr</span>::<span style="color: #00BB00;">filter()</span> masks <span style="color: #0000BB;">stats</span>::filter()</span>
<span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">dplyr</span>::<span style="color: #00BB00;">lag()</span>    masks <span style="color: #0000BB;">stats</span>::lag()</span></code></pre></div>
<div class="section level2">
<h2 id="input-data">Input Data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h2>
<p><code>gscramble</code> operates only for diploid species. The main
functions in <code>gscramble</code> require that genetic data be
provided in R objects (matrices/tibbles/etc.) with particular
formatting. We describe the formats of those objects here. Later we
describe functions for reading data from PLINK format (and perhaps
others, in the future) into these necessary objects for
<code>gscramble</code>.</p>
<p>The main required objects (loosely aligning with .bed/.fam/.bim file
structure of binary PLINK files) are:</p>
<ol style="list-style-type: decimal">
<li>A matrix of genotype data. (<code>Geno</code>)</li>
<li>A tibble of meta data about the individuals in the genotype data
(<code>I_meta</code>)</li>
<li>A tibble of meta data about the markers in the genotype data
(<code>M_meta</code>)</li>
</ol>
<p>On top of those, if you want to simulate hybrid individuals according
to a pedigree, you will need to provide</p>
<ol start="4" style="list-style-type: decimal">
<li>A tibble that describes the rates of genetic recombination along the
chromosomes. (<code>RecRates</code>)</li>
<li>A tibble that describes a pedigree of interest within which sections
of chromosomes are segregated down successive generations, which we call
a “genome simulation pedigree” (<code>GSP</code>)</li>
<li>A tibble that translates the population specifiers in the pedigree
(i.e., <code>GSP</code>) with the population/group labels associated
with the genetic data (i.e., <code>I_meta</code>) for different
hybridization replicates (<code>RepPop</code>)</li>
</ol>
<p>Examples of these objects (which have been heavily downsampled to
make them small enough to host on CRAN, illustrating 78 individuals
genotyped at 100 loci distributed across 3 chromosomes [chr 12, 17,
&amp; 18] and sampled from 4 groups [Pop10, 2, 3, 4]) from a data set of
invasive feral swine (Smyser et al. 2020; full dataset: <a href="https://datadryad.org/stash/dataset/doi:10.5061%2Fdryad.jsxksn05z" class="external-link uri">https://datadryad.org/stash/dataset/doi:10.5061%2Fdryad.jsxksn05z</a>)
are included as data objects with the package. The following sections
discuss each of these data inputs in detail.</p>
<div class="section level3">
<h3 id="matrix-of-genotype-data--geno">Matrix of genotype data. <code>Geno</code><a class="anchor" aria-label="anchor" href="#matrix-of-genotype-data--geno"></a>
</h3>
<p>The genotype data must be provided in a matrix. If there are <span class="math inline">\(N\)</span> individuals and <span class="math inline">\(L\)</span> loci in the diploid species, this
matrix has <span class="math inline">\(N\)</span> rows and <span class="math inline">\(2L\)</span> columns. Each locus gets two adjacent
columns (one allele in each column) and each individual gets one row.
For example, in the first row and first column <code>Geno[1,1]</code> is
the first allele for the first locus for the first individual. In the
first row and second column <code>Geno[1,2]</code> is the second allele
for the first locus for the first individual. In the first row and third
column <code>Geno[1,3]</code> is the first allele at the second locus
for the first individual, and so on.</p>
<p>Here is an example of data on the first 6 loci in the first 4
individuals from the package data <code>Geno</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Geno</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">12</span><span class="op">]</span>
<span class="co">#&gt;      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12]</span>
<span class="co">#&gt; [1,] "G"  "G"  "C"  "C"  "T"  "T"  "A"  "A"  "G"  "G"   "T"   "T"  </span>
<span class="co">#&gt; [2,] "G"  "G"  "C"  "C"  "T"  "T"  "A"  "A"  "G"  "G"   "T"   "T"  </span>
<span class="co">#&gt; [3,] "G"  "G"  "C"  "C"  "T"  "T"  "A"  "A"  "G"  "G"   "T"   "T"  </span>
<span class="co">#&gt; [4,] "G"  "G"  "C"  "C"  "T"  "T"  "A"  "A"  "G"  "G"   "T"   "T"</span></code></pre></div>
<p>These data must be stored as a <strong>character Matrix</strong>.
Don’t pass in a matrix of integers. The alleles can be any characters
whatsoever. This allows the data to be microsatellites,
(<code>"112"</code>, <code>"116"</code>, etc), or microhaplotypes
(<code>"ACCGA"</code>, <code>"ACCTC"</code>, etc.), or SNPs
(<code>"A"</code>, <code>"C"</code>, <code>"G"</code>,
<code>"T"</code>), etc. If you do have a matrix of integers, named, for
example <code>IntMat</code>, you can coerce all the elements of that
matrix to be characters without losing the matrix shape of the data, by
doing this:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/mode.html" class="external-link">storage.mode</a></span><span class="op">(</span><span class="va">IntMat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"character"</span></code></pre></div>
<p>Missing genotype data must be denoted by the standard R missing data
specifier, <code>NA</code>. Don’t go around calling missing data
<code>"-1"</code> and expect it to work properly! Change those
<code>"-1"</code>’s to <code>NA</code>s, or they will be regarded as an
allelic type, rather than as missing data.If you do have missing data,
let’s say they are denoted as <code>"-1"</code>, then you can change all
missing values to <code>NA</code>s using the following:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Geno</span><span class="op">[</span><span class="va">Geno</span> <span class="op">==</span> <span class="st">"-1"</span><span class="op">]</span><span class="op">&lt;-</span><span class="cn">NA</span></code></pre></div>
<p>Since there are no <code>"-1"</code> or <code>NA</code>s in the
<code>Geno</code> dataset, nothing happens.</p>
</div>
<div class="section level3">
<h3 id="individual-meta-data--i_meta">Individual meta data. <code>I_meta</code><a class="anchor" aria-label="anchor" href="#individual-meta-data--i_meta"></a>
</h3>
<p>This is a tibble that gives information about the individuals whose
genotypes are in <code>Geno</code>. This can have many different columns
if you want to have them, but there must be at least two columns:</p>
<ul>
<li>
<code>group</code>: a column that gives the character name of the
group/cluster/population that each individual is considered to be a part
of.</li>
<li>
<code>indiv</code>: a column that gives the character ID of each
individual.</li>
</ul>
<p>The number of rows of this tibble should be exactly equal to the
number of rows on <code>Geno</code> <em>and</em> the order of
individuals in <code>I_meta</code> must correspond exactly to the order
of individuals in <code>Geno</code>.</p>
<p>Here is what the first few rows of the example data’s
<code>I_meta</code> look like:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">I_meta</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 2</span></span>
<span class="co">#&gt;   group indiv    </span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Pop2  ID0033692</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Pop2  ID0014942</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Pop2  ID0014961</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Pop2  ID0016971</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">5</span> Pop2  ID0016972</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">6</span> Pop2  ID0017013</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="marker-meta-data--m_meta">Marker meta data. <code>M_meta</code><a class="anchor" aria-label="anchor" href="#marker-meta-data--m_meta"></a>
</h3>
<p>This input is a tibble of information about the markers in the
<code>Geno</code> matrix. It can have a variety of columns in it, but it
is required to have three:</p>
<ul>
<li>
<code>chrom</code>: the character name of the chromosome upon which
the marker occurs. For example, <code>"1"</code>, <code>"X"</code> of
<code>"Omy28"</code>, or as illustrated in <code>M_meta</code> chr12,
chr17, and chr18.<br><strong>Importantly</strong>, if you are simulating physical linkage
with recombination, then the names of the chromosomes in this file
<em>must</em> correspond <em>exactly</em> to the names of the
chromosomes in <code>RecRates</code> (see next section).</li>
<li>
<code>pos</code>: a numeric (integer or double) column giving the
position of the marker (typically in base pairs, but it could be in
arbitrary units that correspond to position units in
<code>RecRates</code>), along the chromosome. These position values must
be greater than 0.</li>
<li>
<code>variant_id</code>: a character vector of unique ID names for
the markers. These should be globally unique, i.e., don’t have two
markers on different chromosomes that are named the same thing.</li>
</ul>
<p>There must be exactly half as many rows in <code>M_meta</code> as
there are columns in <code>Geno</code>, and the order of markers in
<code>M_meta</code> must correspond exactly to the order of markers in
the columns in <code>Geno</code>.</p>
<p>Here are the first few rows of the example data
<code>M_meta</code>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">M_meta</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 3</span></span>
<span class="co">#&gt;   chrom      pos variant_id         </span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 12     4<span style="text-decoration: underline;">469</span>057 WU_10.2_12_4469057 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 12     5<span style="text-decoration: underline;">238</span>225 ALGA0064411        </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 12     7<span style="text-decoration: underline;">394</span>362 WU_10.2_12_7394362 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 12     7<span style="text-decoration: underline;">651</span>064 ASGA0090707        </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 12     8<span style="text-decoration: underline;">971</span>475 WU_10.2_12_8971475 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 12    11<span style="text-decoration: underline;">034</span>660 WU_10.2_12_11034660</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="recombination-rates--recrates">Recombination rates. <code>RecRates</code><a class="anchor" aria-label="anchor" href="#recombination-rates--recrates"></a>
</h3>
<p>This is a tibble that gives information about the rate of
recombinations in the genome. This is necessary if simulating linked
markers. It is not required in <code>gscramble</code> to know a
crossover rate nor a recombination rate between every adjacent pair of
markers (though if you have that information, you can provide it in
<code>RecRates</code>, [see below]). Rather, the rate of recombination
can be specified in terms of the per-meiosis probability of
recombination in a number of (preferably relatively short—for example,
one megabase or less) bins. <code>RecRates</code> is a tibble which is
required to have four columns:</p>
<ul>
<li>
<code>chrom</code>: the chromosome on which the bin occurs. Note
that the chromosome nomenclature must match exactly that used in
<code>M_meta</code>.</li>
<li>
<code>chrom_len</code>: the length of the chromosome. Yes, the value
for each chromosome will typically be duplicated on each row occupied by
that <code>chrom</code>.</li>
<li>
<code>start_pos</code>: the starting position of the bin. This will
typically be a position along the chromosome in base pairs, though this
position can be in some other units, so long as it corresponds to the
position used in <code>M_meta</code>.</li>
<li>
<code>end_pos</code>: the ending position of the bin.</li>
<li>
<code>rec_prob</code>: the per-meiosis probability of a
recombination occurring in the bin.</li>
</ul>
<p>There are some important notes:</p>
<ol style="list-style-type: lower-roman">
<li>The start point of a bin should be 1 greater than the end point of
the preceding bin.</li>
<li>The positions of all the markers (in <code>M_meta</code>) should be
included amongst the bin intervals defined by <code>start_pos</code> and
<code>end_pos</code>. Most crucially in this regard, the smallest
<code>start_pos</code> should be less than the smallest <code>pos</code>
in <code>M_meta</code> and the greatest <code>end_pos</code> should be
greater than or equal to the largest <code>pos</code> in
<code>M_meta</code>. Otherwise, recombination might never be possible
between some pairs of markers in the data set.</li>
<li>The chromosome length in <code>chrom_len</code>
<strong>must</strong> exceed the position of every marker on the
chromosome in <code>M_meta</code>. If this is not the case then it would
be possible that some markers would be dropped from the data set,
possibly with unexpected or bad results. You can check for errors in
this regard using the function <code><a href="../reference/check_chrom_lengths.html">check_chrom_lengths()</a></code>.</li>
</ol>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">RecRates</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 5</span></span>
<span class="co">#&gt;   chrom chrom_len start_pos end_pos rec_prob</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 12     63<span style="text-decoration: underline;">582</span>536         1   <span style="text-decoration: underline;">79</span>501 0.000<span style="text-decoration: underline;">694</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 12     63<span style="text-decoration: underline;">582</span>536     <span style="text-decoration: underline;">79</span>502 1<span style="text-decoration: underline;">573</span>908 0.001<span style="text-decoration: underline;">31</span> </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 12     63<span style="text-decoration: underline;">582</span>536   1<span style="text-decoration: underline;">573</span>909 2<span style="text-decoration: underline;">097</span>503 0.000<span style="text-decoration: underline;">554</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 12     63<span style="text-decoration: underline;">582</span>536   2<span style="text-decoration: underline;">097</span>504 3<span style="text-decoration: underline;">308</span>663 0.005<span style="text-decoration: underline;">57</span> </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 12     63<span style="text-decoration: underline;">582</span>536   3<span style="text-decoration: underline;">308</span>664 4<span style="text-decoration: underline;">021</span>031 0.017<span style="text-decoration: underline;">9</span>  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 12     63<span style="text-decoration: underline;">582</span>536   4<span style="text-decoration: underline;">021</span>032 5<span style="text-decoration: underline;">026</span>780 0.011<span style="text-decoration: underline;">5</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="genome-simulation-pedigree--gsp">Genome Simulation Pedigree. <code>GSP</code><a class="anchor" aria-label="anchor" href="#genome-simulation-pedigree--gsp"></a>
</h3>
<p>The <code>GSP</code> is a specification of a pedigree within which
sections of chromosome will get segregated across successive generations
represented within the pedigree and is used to to guide sampling of
these chromosomal sections without replacement. This will allow us to
characterize/simulate hybrid individuals for the purpose of assessing
the power to classify individuals to hybrid classes of interest (i.e.,
F1s, F1BC1s, F2BC1s, etc.).</p>
<p>The <code>GSP</code> must be a tibble in which each individual has a
numeric identifier (from 1 up to the number of individuals in the
pedigree). Founders are listed directly (1…nFounders) whereas numbers
can represent more than one individual in successive generations (e.g.,
two F1 individuals sampled within the most basic pedigree illustrated
below are characterized as ‘3’). The founders’ parents are listed as NA
and non-founders have parents listed. Founders’ haplotypes must have
unique IDs and must originate from a specified population (typically
given in capital letters.) Individuals can be sampled from individuals
in the pedigree, and it is up to the user to indicate how many gametes
must be segregated to each individual (from each of its two parents) in
order to consume all the genetic material present among the
founders.</p>
<p>This probably sounds a little abstract, and, indeed, it is.
<code>gscramble</code> includes 15 basic GSPs (illustrated within the
object <code>GSP_opts</code>) that can be retreived by specifying the
types of individuals of interest with the function
<code><a href="../reference/create_GSP.html">create_GSP()</a></code>. As we will illustrate below, you are not
restricted to these GSP configurations. These 15 specific GSP
configurations were of interest to the authors and are included to
illustrate how to structure GSP tibbles that may be of interest to
you.</p>
<p>To try to explain the GSP concept more fully, we will first draw a
picture of the most basic pedigree (the simulation of F1s created
through the hybridization of individuals from two population). We then
show the tibble representation of that same GSP and discuss the
structure of a GSP in light of those two pieces of information.
<img src="../reference/figures/F1-500.png" width="50%" style="display: block; margin: auto;"></p>
<p>Now, for example, let’s say you are interested in simulating F1
hybrids for all pairwise combinations of populations (with the example
data Pop1, Pop2, Pop3, Pop4). F1s are symmetrical such that a Pop1-Pop2
hybrid is identical to a Pop2-Pop1 hybrid. Accordingly, we need to
specify the <code>RepPop</code> tibble to define which populations to
use in each replicate of the simulation:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Pattern</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Pop1"</span>, <span class="st">"Pop2"</span>, <span class="st">"Pop1"</span>, <span class="st">"Pop3"</span>, <span class="st">"Pop1"</span>, <span class="st">"Pop4"</span>, <span class="st">"Pop2"</span>, <span class="st">"Pop3"</span>, <span class="st">"Pop2"</span>, <span class="st">"Pop4"</span>, <span class="st">"Pop3"</span>, <span class="st">"Pop1"</span><span class="op">)</span>
<span class="va">RepPopF1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>rep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, times <span class="op">=</span> <span class="fl">1</span>, each <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>,
                                 pop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p1"</span>, <span class="st">"p2"</span><span class="op">)</span>, times <span class="op">=</span> <span class="fl">6</span><span class="op">)</span>, group <span class="op">=</span> <span class="va">Pattern</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">RepPopF1</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 3</span></span>
<span class="co">#&gt;     rep pop   group</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     1 p1    Pop1 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span>     1 p2    Pop2 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span>     2 p1    Pop1 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span>     2 p2    Pop3 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">5</span>     3 p1    Pop1 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">6</span>     3 p2    Pop4</span></code></pre></div>
<p>Next, we need to define the GSP in which we are using ‘p1’ and ‘p2’
in <code>RepPop</code> to represent Pop1 and Pop2.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gspF1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_GSP.html">create_GSP</a></span><span class="op">(</span>pop1 <span class="op">=</span> <span class="st">"p1"</span>, pop2 <span class="op">=</span> <span class="st">"p2"</span>, F1 <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="va">gspF1</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 11</span></span>
<span class="co">#&gt;     ind  par1  par2 ipar1 ipar2 hap1  hap2  hpop1 hpop2 sample osample</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     1    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span> 1a    1b    p1    p1    <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span>     2    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span> 2a    2b    p2    p2    <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span>     3     1     2     2     2 <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    s3           2</span></code></pre></div>
<p>Building upon this simplistic GSP, let’s build something a little
more ‘complex’ with F1s, F1BC1s, and F1BC2s using the built-in function
<code><a href="../reference/create_GSP.html">create_GSP()</a></code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gspComplex</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_GSP.html">create_GSP</a></span><span class="op">(</span>pop1 <span class="op">=</span> <span class="st">"p1"</span>, pop2 <span class="op">=</span> <span class="st">"p2"</span>, F1 <span class="op">=</span> <span class="cn">TRUE</span>, F1B <span class="op">=</span> <span class="cn">TRUE</span>, F1B2 <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="va">gspComplex</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 7 × 11</span></span>
<span class="co">#&gt;     ind  par1  par2 ipar1 ipar2 hap1  hap2  hpop1 hpop2 sample osample</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     1    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span> 1a    1b    p1    p1    <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span>     2    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span> 2a    2b    p1    p1    <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span>     3    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span> 3a    3b    p1    p1    <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span>     4    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span> 4a    4b    p2    p2    <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">5</span>     5     1     4     2     2 <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    s5           1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">6</span>     6     2     5     2     2 <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    s6           1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">7</span>     7     3     6     2     2 <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    s7           2</span></code></pre></div>
<p>For this example, Pop1 is our population of interest. We are
simulating Pop1-Pop2 F1s, then pairing those F1 individuals with Pop1 to
generate backcrossed hybrids, which are different than Pop1-Pop2,
backcrossed to Pop2 hybrids. Here is a picture of what that looks
like:</p>
<p><img src="../reference/figures/13.svg" width="100%" style="display: block; margin: auto;"></p>
<p>Notice that backcrossed individuals are not symmetrical.</p>
<p>Now say we can simulate hybrid individuals between Pop1 and all other
populations using the <code>RepPop</code> file.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Pattern</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Pop1"</span>, <span class="st">"Pop2"</span>, <span class="st">"Pop1"</span>, <span class="st">"Pop3"</span>, <span class="st">"Pop1"</span>, <span class="st">"Pop4"</span><span class="op">)</span>
<span class="va">RepPopComplex</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>rep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, times <span class="op">=</span> <span class="fl">1</span>, each <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>,
                                 pop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p1"</span>, <span class="st">"p2"</span><span class="op">)</span>, times <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>, group <span class="op">=</span> <span class="va">Pattern</span><span class="op">)</span><span class="op">)</span>
<span class="va">RepPopComplex</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 3</span></span>
<span class="co">#&gt;     rep pop   group</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     1 p1    Pop1 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span>     1 p2    Pop2 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span>     2 p1    Pop1 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span>     2 p2    Pop3 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">5</span>     3 p1    Pop1 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">6</span>     3 p2    Pop4</span></code></pre></div>
<p>If we are interested in Pop1-Pop2, backcrossed to Pop2 hybrids we can
use a <code>RepPop</code> file that looks like this:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Pattern</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Pop2"</span>, <span class="st">"Pop1"</span>, <span class="st">"Pop3"</span>, <span class="st">"Pop1"</span>, <span class="st">"Pop4"</span>, <span class="st">"Pop1"</span><span class="op">)</span>
<span class="va">RepPopComplex2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>rep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, times <span class="op">=</span> <span class="fl">1</span>, each <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>,
                                 pop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p1"</span>, <span class="st">"p2"</span><span class="op">)</span>, times <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>, group <span class="op">=</span> <span class="va">Pattern</span><span class="op">)</span><span class="op">)</span>
<span class="va">RepPopComplex2</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 3</span></span>
<span class="co">#&gt;     rep pop   group</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     1 p1    Pop2 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span>     1 p2    Pop1 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span>     2 p1    Pop3 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span>     2 p2    Pop1 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">5</span>     3 p1    Pop4 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">6</span>     3 p2    Pop1</span></code></pre></div>
<p>We will go over more examples of how to define populations in
<code>RepPop</code> in the section “Mapping populations/collections to
founding populations. <code>RepPop</code>” of this tutorial.</p>
<p>In our previous examples, we specified an initial suite of GSPs that
were of interest, combining various configurations of F1s, F2s, F1BC1s,
and F1BC2s using the <code><a href="../reference/create_GSP.html">create_GSP()</a></code> function. However, you
are not limited to these GSPs and may create your own.</p>
<p>Here are two illustrations of more complex patterns of
hybridization:</p>
<p>The package data object <code>GSP</code> shows an example of a genome
simulation pedigree with 13 members. Here is a picture of what it looks
like:</p>
<p><img src="../reference/figures/13-member-ped.svg" width="100%" style="display: block; margin: auto;"></p>
<p>You can produce this type of plot using the <code><a href="../reference/gsp2dot.html">gsp2dot()</a></code>
function in the ‘gscramble’ package, but it requires the installation of
the GraphViz <code>dot</code> software.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">csv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/13-member-ped.csv"</span>, package <span class="op">=</span> <span class="st">"gscramble"</span><span class="op">)</span>
<span class="va">gsp_tib</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="va">csv</span><span class="op">)</span>
<span class="va">paths</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gsp2dot.html">gsp2dot</a></span><span class="op">(</span>g <span class="op">=</span> <span class="va">gsp_tib</span>, path <span class="op">=</span> <span class="st">"images/13-member-ped"</span><span class="op">)</span>
<span class="co"># now, get rid of the dot and png files</span>
<span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.remove</a></span><span class="op">(</span><span class="va">paths</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p>The tibble specification of that same GSP is printed here:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">csv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/13-member-ped.csv"</span>, package <span class="op">=</span> <span class="st">"gscramble"</span><span class="op">)</span>
<span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="va">csv</span><span class="op">)</span>
<span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">13</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">11</span></span>
<span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span>
<span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> ","</span>
<span class="co">#&gt; <span style="color: #BB0000;">chr</span> (5): hap1, hap2, hpop1, hpop2, sample</span>
<span class="co">#&gt; <span style="color: #00BB00;">dbl</span> (6): ind, par1, par2, ipar1, ipar2, osample</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span>
<span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 13 × 11</span></span>
<span class="co">#&gt;      ind  par1  par2 ipar1 ipar2 hap1  hap2  hpop1 hpop2 sample osample</span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>     1    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span> 1a    1b    A     A     <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>     2    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span> 2a    2b    A     A     <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>     3    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span> 3a    3b    B     B     <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>     4    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span> 4a    4b    A     A     <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>     5    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span> 5a    5b    B     B     <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>     6    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span> 6a    6b    B     B     <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>     7     2     3     2     2 <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    s7           1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>     8     4     5     2     2 <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    s8           1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>     9     1     7     1     1 <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span>    10     7     8     1     1 <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    s10          1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">11</span>    11     8     6     1     1 <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">12</span>    12     9    11     2     2 <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    s12          2</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">13</span>    13     1     6     1     1 <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    s13          1</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="a-four-poplation-gsp--gsp4">A four-poplation GSP. <code>gsp4</code><a class="anchor" aria-label="anchor" href="#a-four-poplation-gsp--gsp4"></a>
</h3>
<p>Or, alternatively, let’s say you are interested in patterns of
hybridization that are not restricted to simply two populations.</p>
<p>For illustration, we will have another pedigree that represents an F1
between populations A and B then mating with an F1 from populations C
and D. The tibble is available in the package data object
<code>gsp4</code>, while the CSV file of it is available at:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/gsp4.csv"</span>, package <span class="op">=</span> <span class="st">"gscramble"</span><span class="op">)</span></code></pre></div>
<p>Here is what it looks like:
<img src="../reference/figures/gsp4-700.png" width="70%" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="mapping-populationscollections-to-founding-populations--reppop">Mapping populations/collections to founding populations.
<code>RepPop</code><a class="anchor" aria-label="anchor" href="#mapping-populationscollections-to-founding-populations--reppop"></a>
</h3>
<p>When you create a genomic simulation pedigree, you will typically
denote the populations that the founders come from with short names,
like “A” or “B”. However, the actual populations in your genotype data
set might be different. For example, in our example data in this R
package, we have groups of individuals given in <code>I_meta</code>.
Here we count up how many individuals in each of these
groups/populations/clusters there are:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">I_meta</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">group</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 × 2</span></span>
<span class="co">#&gt;   group     n</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Pop10    20</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Pop2     20</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Pop3     20</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Pop4     18</span></code></pre></div>
<p>You must use a tibble with columns <code>rep</code>,
<code>pop</code>, <code>group</code>, to indicate which of the founding
populations (“A”, “B”, etc.) correspond to the different groups (from
the <code>group</code> column in, for example, <code>I_meta</code>) in
your genotype data set. Because it is quite likely that you might wish
to iterate the segregation procedure multiple times in a single
simulation, you can specify that by doing multiple “reps” (replicates)
of the procedure.</p>
<p>An example might help. Suppose that we wish to do a simulation with
the pedigree in <code>GSP</code> (13 individuals, 6 of which are
founders: 3 from population “A” and 3 from population “B”). For the
first rep, we might want to map “A” to Pop2 and “B” to Pop3, and in the
second rep we might want to map “A” to Pop1, and “B” to Pop 4. (Note, at
this point, genetic material from 3 individuals from each of those
populations will have been “consumed” from each of these populations and
segregated, without replacement, into the samples from the genomic
simulation pedigree.)</p>
<p>The <code>RepPop</code> tibble that would specify this is given in
the package variable <code>RepPop1</code>:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">RepPop1</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 × 3</span></span>
<span class="co">#&gt;     rep pop   group</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     1 A     Pop2 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span>     1 B     Pop3 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span>     2 A     Pop10</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span>     2 B     Pop4</span></code></pre></div>
<p>For another example, imagine that we want to do three replicates
creating the admixed individuals sampled from the genomic permutation
pedigree, <code>gsp4</code>. The RepPop tibble for that might look like
this:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">RepPop4</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 12 × 3</span></span>
<span class="co">#&gt;      rep pop   group</span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>     1 A     Pop2 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>     1 B     Pop3 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>     1 C     Pop4 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>     1 D     Pop10</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>     2 A     Pop2 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>     2 B     Pop3 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>     2 C     Pop4 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>     2 D     Pop10</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>     3 A     Pop2 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span>     3 B     Pop3 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">11</span>     3 C     Pop4 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">12</span>     3 D     Pop10</span></code></pre></div>
<p>Note that this request will consume 1 individual from each of
populations 2, 3, 4, and 10 (which are mapped to A, B, C, and D,
respectively), and will create 4 simulated, admixed, and sampled
individuals.</p>
</div>
</div>
<div class="section level2">
<h2 id="segregating-chunks-of-genome-without-replacement">Segregating Chunks of Genome Without Replacement<a class="anchor" aria-label="anchor" href="#segregating-chunks-of-genome-without-replacement"></a>
</h2>
<p>Now that we have been through all the input data formats, we can use
them to segregate chunks of chromosome. Note that this first part does
not require the genotypes. We are just simulating big chunks of
chromosome dropping through the pedigrees. The function used for this is
<code><a href="../reference/segregate.html">segregate()</a></code>. The needed inputs are:</p>
<ul>
<li>1 or more genomic permuation pedigrees (like <code>GSP</code> or
<code>gsp4</code>)</li>
<li>A RepPop tibble to go with each genomic permutation pedigree</li>
<li>The recombination probabilities, like <code>RecRates</code>.</li>
</ul>
<p>The final wrinkle here is that we have set this up so that you can
specify, in a single simulation, that replicates from multiple genomic
permutation pedigrees can be requested. This is done by passing
<code><a href="../reference/segregate.html">segregate()</a></code> a tibble that has a list column named
<code>gpp</code> that holds the genomic permutation pedigrees, and a
parallel column named <code>reppop</code> that holds the RepPop tibbles
for each of those genomic permutation pedigrees. Let’s see that by an
example, building on the two examples above:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Input_tibble</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>
  gpp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">GSP</span>, <span class="va">gsp4</span><span class="op">)</span>,
  reppop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">RepPop1</span>, <span class="va">RepPop4</span><span class="op">)</span>
<span class="op">)</span>

<span class="co"># here is what that input object looks like:</span>
<span class="va">Input_tibble</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 2</span></span>
<span class="co">#&gt;   gpp                reppop           </span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>           </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> <span style="color: #949494;">&lt;tibble [13 × 11]&gt;</span> <span style="color: #949494;">&lt;tibble [4 × 3]&gt;</span> </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> <span style="color: #949494;">&lt;tibble [7 × 11]&gt;</span>  <span style="color: #949494;">&lt;tibble [12 × 3]&gt;</span></span></code></pre></div>
<p>The segregation requested is then carried out, using the
recombination rates in <code>RecRates</code> like this:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">15</span><span class="op">)</span> <span class="co"># for reproducibility</span>
<span class="va">Segments</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/segregate.html">segregate</a></span><span class="op">(</span>
  request <span class="op">=</span> <span class="va">Input_tibble</span>,
  RR <span class="op">=</span> <span class="va">RecRates</span>
<span class="op">)</span></code></pre></div>
<p>The output from that is a big tibble. Each row represents one segment
of genetic material amongst the sampled individuals from the genomic
permutation pedigrees. Each segment exists in one of the samples
(<code>samp_index</code>) from a sampled individual with a
<code>ped_sample_id</code> on a given <code>gpp</code> (the index giving
the row of the request input tibble) in a given <code>rep</code> within
the individual. Further, it is on one of two gametes
(<code>gamete_index</code>) that segregated into the individual, and it
came from a certain founding population (<code>pop_origin</code>) that
corresponds to the named groups in the genotype file
(<code>group_origin</code>). And, of course, the segment occupies the
space from <code>start</code> to <code>end</code> on a chromosome
<code>chrom</code>. Finally, the index of the founder haplotype on the
given gpp that this segement descended from is given in
<code>rs_founder_haplotype</code> which is short for “rep-specific
founder haplotype”. This final piece of information is crucial for
segregating variation from the individuals in the <code>Geno</code> file
onto these segments. We take that up in the next section, but first we
will look at the results that we have here.</p>
<p>First, here is a listing of the top of the <code>Segments</code> we
produced above.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Segments</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 286 × 14</span></span>
<span class="co">#&gt;    chrom_f   gpp   rep chrom ped_sample_id samp_index gamete_index</span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 12          1     1 12    10                     1            1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 12          1     1 12    10                     1            1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 17          1     1 17    10                     1            1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 17          1     1 17    10                     1            1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 18          1     1 18    10                     1            1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 18          1     1 18    10                     1            1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 18          1     1 18    10                     1            1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 18          1     1 18    10                     1            1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 12          1     1 12    10                     1            2</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 12          1     1 12    10                     1            2</span>
<span class="co">#&gt; <span style="color: #949494;"># … with 276 more rows, and 7 more variables: gamete_segments &lt;list&gt;,</span></span>
<span class="co">#&gt; <span style="color: #949494;">#   pop_origin &lt;chr&gt;, rs_founder_haplo &lt;int&gt;, start &lt;dbl&gt;, end &lt;dbl&gt;,</span></span>
<span class="co">#&gt; <span style="color: #949494;">#   group_origin &lt;chr&gt;, sim_level_founder_haplo &lt;int&gt;</span></span></code></pre></div>
<p>The table is a little wide, so we will show the first half of the
columns and then the second half here.</p>
<p>First half</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Segments</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">gpp</span><span class="op">:</span><span class="va">pop_origin</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 286 × 8</span></span>
<span class="co">#&gt;      gpp   rep chrom ped_sample_id samp_index gamete_index gamete_segments</span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>         </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>     1     1 12    10                     1            1 <span style="color: #949494;">&lt;dbl [3]&gt;</span>      </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>     1     1 12    10                     1            1 <span style="color: #949494;">&lt;dbl [3]&gt;</span>      </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>     1     1 17    10                     1            1 <span style="color: #949494;">&lt;dbl [3]&gt;</span>      </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>     1     1 17    10                     1            1 <span style="color: #949494;">&lt;dbl [3]&gt;</span>      </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>     1     1 18    10                     1            1 <span style="color: #949494;">&lt;dbl [5]&gt;</span>      </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>     1     1 18    10                     1            1 <span style="color: #949494;">&lt;dbl [5]&gt;</span>      </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>     1     1 18    10                     1            1 <span style="color: #949494;">&lt;dbl [5]&gt;</span>      </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>     1     1 18    10                     1            1 <span style="color: #949494;">&lt;dbl [5]&gt;</span>      </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>     1     1 12    10                     1            2 <span style="color: #949494;">&lt;dbl [5]&gt;</span>      </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span>     1     1 12    10                     1            2 <span style="color: #949494;">&lt;dbl [5]&gt;</span>      </span>
<span class="co">#&gt; <span style="color: #949494;"># … with 276 more rows, and 1 more variable: pop_origin &lt;chr&gt;</span></span></code></pre></div>
<p>Second half</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Segments</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">rs_founder_haplo</span><span class="op">:</span><span class="va">group_origin</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 286 × 4</span></span>
<span class="co">#&gt;    rs_founder_haplo     start       end group_origin</span>
<span class="co">#&gt;               <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>                3        0   4<span style="text-decoration: underline;">888</span>302. Pop2        </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>                2  4<span style="text-decoration: underline;">888</span>302. 63<span style="text-decoration: underline;">582</span>536  Pop3        </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>                2        0  64<span style="text-decoration: underline;">743</span>108. Pop3        </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>                4 64<span style="text-decoration: underline;">743</span>108. 69<span style="text-decoration: underline;">302</span>804  Pop2        </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>                4        0   4<span style="text-decoration: underline;">583</span>912. Pop2        </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>                2  4<span style="text-decoration: underline;">583</span>912.  9<span style="text-decoration: underline;">900</span>731. Pop3        </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>                4  9<span style="text-decoration: underline;">900</span>731. 56<span style="text-decoration: underline;">967</span>240. Pop2        </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>                2 56<span style="text-decoration: underline;">967</span>240. 61<span style="text-decoration: underline;">201</span>108  Pop3        </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>                6        0   9<span style="text-decoration: underline;">808</span>646. Pop2        </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span>                4  9<span style="text-decoration: underline;">808</span>646. 50<span style="text-decoration: underline;">453</span>258. Pop3        </span>
<span class="co">#&gt; <span style="color: #949494;"># … with 276 more rows</span></span></code></pre></div>
<div class="section level3">
<h3 id="visualizing-those-chunks-of-genome">Visualizing those chunks of genome<a class="anchor" aria-label="anchor" href="#visualizing-those-chunks-of-genome"></a>
</h3>
<p>We have a convenience function called
<code>plot_simulated_chromosome_segments()</code> that let’s you quickly
visualize the results. Let’s try it here:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_simulated_chromomsome_segments.html">plot_simulated_chromomsome_segments</a></span><span class="op">(</span><span class="va">Segments</span>, <span class="va">RecRates</span><span class="op">)</span>
<span class="va">g</span></code></pre></div>
<p><img src="gscramble-tutorial_1.2_files/figure-html/unnamed-chunk-28-1.png" width="720"></p>
</div>
</div>
<div class="section level2">
<h2 id="simulating-alleles-at-markers-within-segments">Simulating alleles at markers within segments<a class="anchor" aria-label="anchor" href="#simulating-alleles-at-markers-within-segments"></a>
</h2>
<p>There are a few separate steps required to make this happen, and they
all occur within the function <code><a href="../reference/segments2markers.html">segments2markers()</a></code>. The steps
that happen are:</p>
<ol style="list-style-type: decimal">
<li>The genotype matrix is reorganized according to the
<code>group</code> specification in the <code>I_meta</code>, so that
individuals in the same group are adjacent to one another in the
genotype matrix.</li>
<li>The alleles carried by the individuals within each population are
permuted.</li>
<li>The chromosomes carrying these permuted alleles from
<code>Geno</code> are mapped to the chromosomes carried by the founders
in the simulation performed by <code><a href="../reference/segregate.html">segregate()</a></code>.</li>
<li>The alleles on the founder chromosomes are propagated to the
descendant segments in the individuals sampled from the GSP.</li>
<li>Missing data is dealt with. (If a site has one gene copy missing
after the permutation and mapping described above, the other gene copy
must be declared missing, as well).</li>
<li>The output genotypes are formatted for return, along with some
information about the true admixture proportions of each
individual.</li>
</ol>
<p>The input to <code><a href="../reference/segments2markers.html">segments2markers()</a></code> is:</p>
<ul>
<li>A tibble of segments, like that returned from
<code><a href="https://rdrr.io/r/graphics/segments.html" class="external-link">segments()</a></code>. This is the variable <code>Segments</code>
defined above in this vignette example,</li>
<li>The individual meta data,</li>
<li>The marker meta data,</li>
<li>The genotype data.</li>
</ul>
<p>We invoke it like this:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/segments2markers.html">segments2markers</a></span><span class="op">(</span>
  Segs <span class="op">=</span> <span class="va">Segments</span>,
  Im <span class="op">=</span> <span class="va">I_meta</span>,
  Mm <span class="op">=</span> <span class="va">M_meta</span>,
  G <span class="op">=</span> <span class="va">Geno</span>
<span class="op">)</span></code></pre></div>
<p>The output is a list of three components:</p>
<ul>
<li>
<p><code>$ret_geno</code> is an N x 2L matrix of returned genotypes.
The two alleles at each locus are in two adjacent columns.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">Markers</span><span class="op">$</span><span class="va">ret_geno</span><span class="op">)</span>
<span class="co">#&gt; [1]  78 200</span></code></pre></div>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># genotypes of the first 10 individuals at the first 3 markers</span>
<span class="va">Markers</span><span class="op">$</span><span class="va">ret_geno</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>
<span class="co">#&gt;       [,1] [,2] [,3] [,4] [,5] [,6]</span>
<span class="co">#&gt;  [1,] "G"  "G"  "C"  "C"  "T"  "T" </span>
<span class="co">#&gt;  [2,] "G"  "G"  "C"  "C"  "T"  "T" </span>
<span class="co">#&gt;  [3,] "G"  "G"  "C"  "C"  "T"  "T" </span>
<span class="co">#&gt;  [4,] "G"  "G"  "C"  "C"  "T"  "T" </span>
<span class="co">#&gt;  [5,] "G"  "G"  "C"  "C"  "T"  "T" </span>
<span class="co">#&gt;  [6,] "G"  "G"  "C"  "C"  "C"  "T" </span>
<span class="co">#&gt;  [7,] "G"  "G"  "C"  "C"  "T"  "T" </span>
<span class="co">#&gt;  [8,] "G"  "G"  "C"  "C"  "T"  "C" </span>
<span class="co">#&gt;  [9,] "G"  "A"  "C"  "T"  "C"  "C" </span>
<span class="co">#&gt; [10,] "G"  "A"  "C"  "C"  "C"  "C"</span></code></pre></div>
</li>
<li>
<p><code>$ret_ids</code>: a tibble with IDs of the individuals
corresponding to the rows in the <code>ret_geno</code> matrix. There are
two columns: <code>group</code> and <code>indiv</code>. The first rows
include the samples from the GSP. Their group is <code>ped_hybs</code>
(meaning “sampled hybrids from the pedigree”) and their IDs are in the
<code>indiv</code> column in the format of an <code>h-</code> followed
by a string with replaced values of:
<code>gpp-rep-ped_sample_id-samp_index-matrix_row</code>. Those are
(they need explaining):</p>
<ul>
<li>
<code>rep</code>:</li>
<li>
<code>ped</code>:</li>
<li>
<code>sample_id</code>:</li>
<li>
<code>samp_index</code>:</li>
<li>
<code>matrix_row</code>:</li>
</ul>
<p>After the individual hybrid samples from the pedigree, there are rows
of all the remaining individuals whose genotypes after permutation were
not involved as founders in the segregation simulation. These are given
the name <code>permed</code> + their original IDs. The group that each
belongs to is in the group column. Things are organized this way so that
the whole genotype can easily be passed to ADMIXTURE and the admixture
fractions of the hybrid individuals estimated using the
<code>permed_*</code> individuals as samples of known cluster/group
origin.</p>
</li>
<li><p><code>$hyb_Qs</code>: A tibble with the true admixture fraction
(calculated as the total proportion of genome length from each group) of
each simulated hybrid.</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="back-to-plink-format">Back to plink format<a class="anchor" aria-label="anchor" href="#back-to-plink-format"></a>
</h2>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># note, we are only using a temporary file here for the output</span>
<span class="co"># because this is in the vignette of an R package.  You, yourself,</span>
<span class="co"># will likely want to write the results to a directory in your</span>
<span class="co"># home directory somewhere, like `prefix = ~/my_stuff/gscram-sim-1`, etc.</span>
<span class="va">tfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="../reference/gscramble2plink.html">gscramble2plink</a></span><span class="op">(</span>
  I_meta <span class="op">=</span> <span class="va">Markers</span><span class="op">$</span><span class="va">ret_ids</span>,
  M_meta <span class="op">=</span> <span class="va">M_meta</span>,
  Geno <span class="op">=</span> <span class="va">Markers</span><span class="op">$</span><span class="va">ret_geno</span>,
  prefix <span class="op">=</span> <span class="va">tfile</span>
<span class="op">)</span>
<span class="co">#&gt; pedfile written to /var/folders/xg/mz_qt7q54yv_hwzvhskwx2c00000gp/T//RtmpX3FLGV/file112ae46ab53.ped</span>
<span class="co">#&gt; mapfile written to /var/folders/xg/mz_qt7q54yv_hwzvhskwx2c00000gp/T//RtmpX3FLGV/file112ae46ab53.map</span>
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by The package maintainer.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.4.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
