<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ gscramble</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gscramble</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/gscramble-tutorial.html">A first tutorial for `gscramble`</a>
    </li>
    <li>
      <a href="../articles/about-createGSP.html">Pedigrees Available from create_GSP()</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="high-level-simulation-functions">High level simulation functions <a href="#high-level-simulation-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>These are functions that are intended for use by the user when preparing inputs for a gscramble simulation and for conducting those simulation.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="create_GSP.html">create_GSP()</a></code> </p>
        </td>
        <td><p>Create a GSP from user inputs about what type of hybrids from which populations to create hybrids</p></td>
      </tr><tr><td>
          <p><code><a href="segregate.html">segregate()</a></code> </p>
        </td>
        <td><p>Segregate segments down genomic simulation pedigrees</p></td>
      </tr><tr><td>
          <p><code><a href="segments2markers.html">segments2markers()</a></code> </p>
        </td>
        <td><p>Map alleles from scrambled founders to the sampled segments from a GSP.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="functions-for-summarizing-and-visualizing-simulation-results-or-inputs">Functions for summarizing and visualizing simulation results or inputs <a href="#functions-for-summarizing-and-visualizing-simulation-results-or-inputs" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>These are just a few functions to summarize the results from gscramble simulations or to visualize GSPs.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="computeQs_from_segments.html">computeQs_from_segments()</a></code> </p>
        </td>
        <td><p>return the admixture fractions of sampled individuals</p></td>
      </tr><tr><td>
          <p><code><a href="plot_simulated_chromomsome_segments.html">plot_simulated_chromomsome_segments()</a></code> </p>
        </td>
        <td><p>Plot the simulated chromosomes of an individual</p></td>
      </tr><tr><td>
          <p><code><a href="gsp2dot.html">gsp2dot()</a></code> </p>
        </td>
        <td><p>Write a dot file to represent a genome simulation pedigree</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="data-input-and-output-functions">Data input and output functions <a href="#data-input-and-output-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>These functions help to get data in PLINK format into gscramble, and also converting gscramble objects into PLINK files.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plink2gscramble.html">plink2gscramble()</a></code> </p>
        </td>
        <td><p>read plink-formatted .map and .ped files into gscramble format</p></td>
      </tr><tr><td>
          <p><code><a href="gscramble2plink.html">gscramble2plink()</a></code> </p>
        </td>
        <td><p>Write gscramble I_meta, M_meta, and Geno to a plink file</p></td>
      </tr><tr><td>
          <p><code><a href="plink_map2rec_rates.html">plink_map2rec_rates()</a></code> </p>
        </td>
        <td><p>Convert a PLINK map file to gscramble RecRates bins in a tibble</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="functions-for-checking-the-formatting-and-validity-of-inputs">Functions for checking the formatting and validity of inputs <a href="#functions-for-checking-the-formatting-and-validity-of-inputs" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>These functions, some of which are used internally by other functions, can let the user check to see if their inputs are OK.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="check_chrom_lengths.html">check_chrom_lengths()</a></code> </p>
        </td>
        <td><p>check that the chromosome lengths are acceptable given recombination rates</p></td>
      </tr><tr><td>
          <p><code><a href="check_gsp_for_validity_and_saturation.html">check_gsp_for_validity_and_saturation()</a></code> </p>
        </td>
        <td><p>Check that GSP does not reuse genetic material and yet uses all of it</p></td>
      </tr><tr><td>
          <p><code><a href="check_pedigree_for_inbreeding.html">check_pedigree_for_inbreeding()</a></code> </p>
        </td>
        <td><p>Check the a GSP (in prepped list format) for inbreeding loops</p></td>
      </tr><tr><td>
          <p><code><a href="check_reppop.html">check_reppop()</a></code> </p>
        </td>
        <td><p>check that the RepPop is formatted correctly</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="example-data-objects">Example data objects <a href="#example-data-objects" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>These data objects are provided as examples of the inputs to gscramble. They are used in the vignettes as well.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="Geno.html">Geno</a></code> </p>
        </td>
        <td><p>Genotype matrix of 78 individuals and 100 SNP markers</p></td>
      </tr><tr><td>
          <p><code><a href="I_meta.html">I_meta</a></code> </p>
        </td>
        <td><p>Metadata for 78 individuals</p></td>
      </tr><tr><td>
          <p><code><a href="M_meta.html">M_meta</a></code> </p>
        </td>
        <td><p>Metadata for 100 molecular markers</p></td>
      </tr><tr><td>
          <p><code><a href="RecRates.html">RecRates</a></code> </p>
        </td>
        <td><p>Recombination rate data for many roughly 1 Mb bins</p></td>
      </tr><tr><td>
          <p><code><a href="gsp4.html">gsp4</a></code> </p>
        </td>
        <td><p>Tibble holding specification for a 7 member genomic permutation pedigree.</p></td>
      </tr><tr><td>
          <p><code><a href="GSP.html">GSP</a></code> </p>
        </td>
        <td><p>Example Genomic Simulation Pedigree (GSP) with 13 members</p></td>
      </tr><tr><td>
          <p><code><a href="RepPop1.html">RepPop1</a></code> </p>
        </td>
        <td><p>A simple example of a reppop table</p></td>
      </tr><tr><td>
          <p><code><a href="RepPop4.html">RepPop4</a></code> </p>
        </td>
        <td><p>Another simple example of a reppop table</p></td>
      </tr><tr><td>
          <p><code><a href="GSP_opts.html">GSP_opts</a></code> </p>
        </td>
        <td><p>A list of tibbles specifying the pedigrees available from <code>createGSP()</code></p></td>
      </tr><tr><td>
          <p><code><a href="example_segments.html">example_segments</a></code> </p>
        </td>
        <td><p>Example of a segments tibble</p></td>
      </tr><tr><td>
          <p><code><a href="gsp3.html">gsp3</a></code> </p>
        </td>
        <td><p>Tibble holding specification for a 5 member genomic permutation pedigree.</p></td>
      </tr><tr><td>
          <p><code><a href="example_chrom_lengths.html">example_chrom_lengths</a></code> </p>
        </td>
        <td><p>Lengths of the three chromosomes used in the example data set</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="low-level-functions">Low-level functions <a href="#low-level-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>These are functions that are not intended for use by the general user, but they are included in the documentation for the interest of developers and they are exported in the package to help with debugging and development.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="drop_segs_down_gsp.html">drop_segs_down_gsp()</a></code> </p>
        </td>
        <td><p>High level function for dropping segments down a GSP</p></td>
      </tr><tr><td>
          <p><code><a href="make_subscript_matrix.html">make_subscript_matrix()</a></code> </p>
        </td>
        <td><p>makes a two-column matrix for subscripting alleles out of a genotype matrix</p></td>
      </tr><tr><td>
          <p><code><a href="mat_scramble.html">mat_scramble()</a></code> </p>
        </td>
        <td><p>Scramble a matrix of genotype data</p></td>
      </tr><tr><td>
          <p><code><a href="perm_gs_by_pops.html">perm_gs_by_pops()</a></code> </p>
        </td>
        <td><p>Take the output of rearrange_genos and permute everyone by population</p></td>
      </tr><tr><td>
          <p><code><a href="prep_gsp_for_hap_dropping.html">prep_gsp_for_hap_dropping()</a></code> </p>
        </td>
        <td><p>Take a gsp in tibble form and make a list suitable for gene dropping</p></td>
      </tr><tr><td>
          <p><code><a href="rearrange_genos.html">rearrange_genos()</a></code> </p>
        </td>
        <td><p>rearrange genotypes into separate columns for each haplotype.</p></td>
      </tr><tr><td>
          <p><code><a href="recomb_point.html">recomb_point()</a></code> </p>
        </td>
        <td><p>Randomly sample the positions of recombinations on a chromosome</p></td>
      </tr><tr><td>
          <p><code><a href="seg2tib.html">seg2tib()</a></code> </p>
        </td>
        <td><p>Takes a gamete in segment format and returns a tibble with Pop and indiv_index</p></td>
      </tr><tr><td>
          <p><code><a href="seg_haps_through_gsp.html">seg_haps_through_gsp()</a></code> </p>
        </td>
        <td><p>Segregate haplotypes through a genome simulation pedigree list</p></td>
      </tr><tr><td>
          <p><code><a href="sim_level_founder_haplos.html">sim_level_founder_haplos()</a></code> </p>
        </td>
        <td><p>computes the simulation-level founder haplotype index for each founder haplo</p></td>
      </tr><tr><td>
          <p><code><a href="tidy_up_sampled_haplos.html">tidy_up_sampled_haplos()</a></code> </p>
        </td>
        <td><p>takes the waka_waka table and returns a simple tibble with results for each sampled "hybridized" gamete from each chromosome
#'
Not yet documented</p></td>
      </tr><tr><td>
          <p><code><a href="xover.html">xover()</a></code> </p>
        </td>
        <td><p>internal function to do crossovers and create recombinations</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Eric C. Anderson, Rachael M. Giglio, Timothy J. Smyser.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

