<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Segregate segments down genomic simulation pedigrees — segregate • gscramble</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Segregate segments down genomic simulation pedigrees — segregate"><meta property="og:description" content="Given a collection of genomic simulation pedigrees and requests
for how many simulations should be done (in the request input),
as well as recombination rates, this simulates the segregation
of segments down through the pedigrees"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gscramble</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/gscramble-tutorial.html">A first tutorial for `gscramble`</a>
    </li>
    <li>
      <a href="../articles/about-createGSP.html">Pedigrees Available from create_GSP()</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Segregate segments down genomic simulation pedigrees</h1>
    
    <div class="hidden name"><code>segregate.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Given a collection of genomic simulation pedigrees and requests
for how many simulations should be done (in the <code>request</code> input),
as well as recombination rates, this simulates the segregation
of segments down through the pedigrees</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">segregate</span><span class="op">(</span><span class="va">request</span>, <span class="va">RR</span>, MM <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>request</dt>
<dd><p>a tibble with list columns "gpp" and "reppop".
Each element of the gpp column is a tibble giving a genomic simulation
pedigree as documented as the input for <code><a href="prep_gsp_for_hap_dropping.html">prep_gsp_for_hap_dropping()</a></code>.
Each element of the "reppop" column is a tibble with columns
<code>index</code>, <code>pop</code>, <code>group</code>, to indicate which of the founding
populations ("A", "B", etc.) correspond to the different groups
(from the <code>group</code> column in, for example, the meta data for individuals
in your genotype data set, like the data object <code>I_meta</code>).
Because it is quite likely that you might wish to iterate
the segregation procedure multiple
times in a single simulation, you can specify that by doing multiple
"reps" (replicates) of the procedure.  <strong>BIG NOTE</strong>: The values in the index column that you
choose must start at 1 and should be dense within.  In other words, if
the max value in the index column is N, then every integer from 1 to N
must be in there.</p></dd>


<dt>RR</dt>
<dd><p>the recombination rates in the format of the package data</p></dd>


<dt>MM</dt>
<dd><p>the marker meta data tibble (like M_meta).  If this is NULL (the default) that
is fine.  If not, then it uses the order of the markers in MM to define
the levels of a chrom_f column so that we can sort the rows of the output
correctly, with respect to markers in the Genotype data frame.  This will
let us more efficiently subscript the markers out of the matrix. If MM is
not present, then the function will create <code>chrom_f</code> by using the
order of the chromosomes from RR. If MM is not NULL, then the function
will also check to make sure that the markers are within the extent of the
recombination rate bins, giving an error otherwise.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>The output from this function is a tibble.  Each row represents one segment of genetic
material amongst the sampled individuals from the genomic permutation pedigrees. The columns give
information about the provenance and destination of that segment as follows.
Each segment exists in one of the samples (<code>samp_index</code>) from a sampled individual
with a <code>ped_sample_id</code> on a given <code>gpp</code> (the index giving the row of the request input tibble)
in a given <code>index</code> within the individual.  Further, it is on one of two gametes
(<code>gamete_index</code>) that segregated into the individual, and it came from a certain founding
population (<code>pop_origin</code>) that corresponds to the named groups in the genotype file (<code>group_origin</code>).
And, of course, the segment occupies the space from <code>start</code> to <code>end</code> on a chromosome <code>chrom</code>.
Finally, the index of the founder haplotype on the given gpp that this segement descended from is
given in <code>rs_founder_haplotype</code> which is short for "rep-specific founder haplotype". This final
piece of information is crucial for segregating variation from the individuals in the <code>Geno</code> file
onto these segments. Finally, the column <code>sim_level_founder_haplo</code> assigns a unique index for each founder
haplotype. This is necessary because any simulation can involve multiple gpps and/or indexes of gpps,
and the founders in each of those must all be unique within a simulation. so that those haplotypes
can all, eventually, be accessed easily out of the genotype matrix.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="co"># We construct an example here where we will request segregation</span></span>
<span class="r-in"><span class="co"># down a GSP with two F1s and F1B backcrosses between two hypothetical</span></span>
<span class="r-in"><span class="co"># populations, A and B.</span></span>
<span class="r-in"><span class="va">gsp_f1f1b</span> <span class="op">&lt;-</span> <span class="fu"><a href="create_GSP.html">create_GSP</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, F1 <span class="op">=</span> <span class="cn">TRUE</span>, F1B <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># We will imagine that in our marker data there are three groups</span></span>
<span class="r-in"><span class="co"># labelled "grp1", "grp2", and "grp3", and we want to create the F1Bs with backcrossing</span></span>
<span class="r-in"><span class="co"># only to grp3.</span></span>
<span class="r-in"><span class="va">reppop</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span class="r-in">    index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span class="r-in">    pop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>,</span>
<span class="r-in">    group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grp3"</span>, <span class="st">"grp1"</span>, <span class="st">"grp3"</span>, <span class="st">"grp2"</span><span class="op">)</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># combine those into a request</span></span>
<span class="r-in"><span class="va">request</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span class="r-in">   gpp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">gsp_f1f1b</span><span class="op">)</span>,</span>
<span class="r-in">   reppop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">reppop</span><span class="op">)</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"></span>
<span class="r-in"><span class="va">result1</span> <span class="op">&lt;-</span> <span class="fu">segregate</span><span class="op">(</span><span class="va">request</span>, <span class="va">RecRates</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># here we pass it some markers, too</span></span>
<span class="r-in"><span class="va">result2</span> <span class="op">&lt;-</span> <span class="fu">segregate</span><span class="op">(</span><span class="va">request</span>, <span class="va">RecRates</span>, <span class="va">M_meta</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="va">result1</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 52 × 14</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    chrom_f   gpp index chrom ped_sample_id samp_index gamete_index</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> 12          1     1 12    4                      1            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> 17          1     1 17    4                      1            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> 18          1     1 18    4                      1            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> 12          1     1 12    4                      1            2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> 17          1     1 17    4                      1            2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> 18          1     1 18    4                      1            2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> 12          1     1 12    5                      1            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> 17          1     1 17    5                      1            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> 18          1     1 18    5                      1            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> 12          1     1 12    5                      1            2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># … with 42 more rows, and 7 more variables: gamete_segments &lt;list&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   pop_origin &lt;chr&gt;, rs_founder_haplo &lt;int&gt;, start &lt;dbl&gt;, end &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   group_origin &lt;chr&gt;, sim_level_founder_haplo &lt;int&gt;</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="va">result2</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 48 × 14</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    chrom_f   gpp index chrom ped_sample_id samp_index gamete_index</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> 12          1     1 12    4                      1            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> 17          1     1 17    4                      1            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> 18          1     1 18    4                      1            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> 12          1     1 12    4                      1            2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> 17          1     1 17    4                      1            2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> 18          1     1 18    4                      1            2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> 12          1     1 12    5                      1            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> 17          1     1 17    5                      1            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> 18          1     1 18    5                      1            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> 12          1     1 12    5                      1            2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># … with 38 more rows, and 7 more variables: gamete_segments &lt;list&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   pop_origin &lt;chr&gt;, rs_founder_haplo &lt;int&gt;, start &lt;dbl&gt;, end &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   group_origin &lt;chr&gt;, sim_level_founder_haplo &lt;int&gt;</span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Eric C. Anderson, Rachael M. Giglio, Timothy J. Smyser.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

